File: backend/internal/api/portal_branding.go
// Package api — validations et encodage du branding du portail captif,
// PARTAGÉS entre les réglages du compte (PUT /api/settings, N°45 → N°139) et
// les surcharges SITE/ROUTEUR de la chaîne de personnalisation N°182
// (POST/PUT /api/sites, PUT /api/routers/{id}/portal). Mêmes bornes, mêmes
// messages d'erreur aux trois niveaux : le contrat du gérant ne dépend pas du
// niveau où il personnalise. Tout JSON final est RE-SÉRIALISÉ côté serveur —
// le client ne peut rien injecter d'autre que ces champs validés.
package api

import (
	"encoding/json"
	"fmt"
	"strings"

	"mikcloud/hotspot-api/internal/hotpage"
	"mikcloud/hotspot-api/internal/model"
)

// portalOverrideReq — corps de requête d'une surcharge de branding (site ou
// routeur). SÉMANTIQUE : la surcharge ENTIÈRE est remplacée à chaque écriture
// (la console renvoie l'état complet du formulaire, comme la carte Expérience
// du compte — pas de nil = inchangé). Champs vides = le niveau supérieur
// s'applique (« vide = hérite », N°182).
type portalOverrideReq struct {
	DisplayName *string             `json:"displayName"`
	LogoURL     *string             `json:"logoUrl"`
	BannerURL   *string             `json:"bannerUrl"`
	WaveLink    *string             `json:"waveLink"`
	Style       *string             `json:"portalStyle"`
	Welcome     *string             `json:"portalWelcome"`
	Promos      *[]portalPromoReq   `json:"portalPromos"`
	Socials     *[]portalSocialReq  `json:"portalSocials"`
	Services    *[]portalServiceReq `json:"portalServices"`
	Slides      *[]string           `json:"portalSlides"`
	Ticker      *[]string           `json:"portalTicker"`
	Whatsapp    *portalWhatsappReq  `json:"portalWhatsapp"`
}

// portalValidateLogo — data:image/ ≤ 300 Ko (même contrat que le logo du
// compte, N°135). Vide = pas de surcharge du logo.
func portalValidateLogo(v string) error {
	v = strings.TrimSpace(v)
	if v == "" {
		return nil
	}
	if !strings.HasPrefix(v, "data:image/") {
		return fmt.Errorf("Logo invalide : image intégrée (data:image/…) requise")
	}
	if len(v) > 300*1024 {
		return fmt.Errorf("Logo trop volumineux (300 Ko max)")
	}
	return nil
}

// portalValidateBanner — data:image/ ≤ 500 Ko OU URL https:// (N°45). Vide =
// pas de surcharge de la bannière.
func portalValidateBanner(v string) error {
	v = strings.TrimSpace(v)
	if v == "" {
		return nil
	}
	if !strings.HasPrefix(v, "data:image/") && !strings.HasPrefix(v, "https://") {
		return fmt.Errorf("Bannière invalide : image intégrée (data:image/…) ou URL https:// requise")
	}
	if strings.HasPrefix(v, "data:image/") && len(v) > 500*1024 {
		return fmt.Errorf("Bannière trop volumineuse (500 Ko max)")
	}
	return nil
}

// portalValidateStyle — "" (hérite) | commercial | hospitality (N°55).
func portalValidateStyle(v string) error {
	v = strings.TrimSpace(v)
	if v != "" && v != "commercial" && v != "hospitality" {
		return fmt.Errorf("Style de portail invalide (commercial ou hospitality)")
	}
	return nil
}

// portalValidateWelcome — ≤ 200 caractères (N°55).
func portalValidateWelcome(v string) error {
	if len(strings.TrimSpace(v)) > 200 {
		return fmt.Errorf("Message de bienvenue trop long (200 caractères max)")
	}
	return nil
}

// portalEncodePromos — vitrine hospitalité : ≤ 6 lignes, titres 1-60,
// descriptions ≤ 160, images https ≤ 300 car., prix ≤ 30, lien https ≤ 300,
// ID conservé s'il est bien formé sinon généré (N°56 — les compteurs
// analytics ne repartent jamais de zéro au round-trip console GET→PUT).
func portalEncodePromos(list []portalPromoReq) (string, error) {
	if len(list) > 6 {
		return "", fmt.Errorf("au plus 6 promos")
	}
	type promo struct {
		ID         string `json:"id,omitempty"`
		Title      string `json:"title"`
		Desc       string `json:"desc,omitempty"`
		ImageURL   string `json:"imageUrl,omitempty"`
		PriceLabel string `json:"priceLabel,omitempty"`
		Link       string `json:"link,omitempty"`
	}
	out := make([]promo, 0, len(list))
	for _, it := range list {
		title := strings.TrimSpace(it.Title)
		if title == "" || len(title) > 60 {
			return "", fmt.Errorf("titre de promo requis (1-60 caractères)")
		}
		desc := strings.TrimSpace(it.Desc)
		if len(desc) > 160 {
			return "", fmt.Errorf("description de promo trop longue (160 caractères max)")
		}
		img := strings.TrimSpace(it.ImageURL)
		if img != "" && !strings.HasPrefix(img, "https://") {
			return "", fmt.Errorf("image de promo invalide : URL https:// requise")
		}
		if len(img) > 300 {
			return "", fmt.Errorf("URL d'image trop longue")
		}
		price := strings.TrimSpace(it.PriceLabel)
		if len(price) > 30 {
			return "", fmt.Errorf("prix trop long (30 caractères max)")
		}
		id := strings.TrimSpace(it.ID)
		if id != "" && !promoIDValid(id) {
			return "", fmt.Errorf("id de promo invalide")
		}
		if id == "" {
			id = model.NewID("p")
		}
		link := strings.TrimSpace(it.Link)
		if link != "" && (!strings.HasPrefix(link, "https://") || len(link) > 300) {
			return "", fmt.Errorf("lien de promo invalide (https://, 300 caractères max)")
		}
		out = append(out, promo{ID: id, Title: title, Desc: desc, ImageURL: img, PriceLabel: price, Link: link})
	}
	if len(out) == 0 {
		return "", nil // liste vidée = promos retirées
	}
	b, err := json.Marshal(out)
	if err != nil {
		return "", err
	}
	return string(b), nil
}

// portalEncodeSocials — liens réseaux sociaux du mode hospitalité : ≤ 4,
// libellé 1-30, URL https ≤ 200.
func portalEncodeSocials(list []portalSocialReq) (string, error) {
	if len(list) > 4 {
		return "", fmt.Errorf("au plus 4 liens sociaux")
	}
	type social struct {
		Label string `json:"label"`
		URL   string `json:"url"`
	}
	out := make([]social, 0, len(list))
	for _, it := range list {
		label := strings.TrimSpace(it.Label)
		url := strings.TrimSpace(it.URL)
		if label == "" || len(label) > 30 {
			return "", fmt.Errorf("libellé de lien requis (1-30 caractères)")
		}
		if !strings.HasPrefix(url, "https://") || len(url) > 200 {
			return "", fmt.Errorf("URL de lien invalide (https://, 200 caractères max)")
		}
		out = append(out, social{Label: label, URL: url})
	}
	if len(out) == 0 {
		return "", nil
	}
	b, err := json.Marshal(out)
	if err != nil {
		return "", err
	}
	return string(b), nil
}

// portalEncodeServices — section « Nos Services » (N°137) : ≤ 6 lignes,
// libellé 1-60, icône dans la whitelist curée (défaut fa-wifi).
func portalEncodeServices(list []portalServiceReq) (string, error) {
	if len(list) > 6 {
		return "", fmt.Errorf("au plus 6 services")
	}
	type service struct {
		Icon  string `json:"icon,omitempty"`
		Label string `json:"label"`
	}
	out := make([]service, 0, len(list))
	for _, it := range list {
		label := strings.TrimSpace(it.Label)
		if label == "" || len(label) > 60 {
			return "", fmt.Errorf("nom de service requis (1-60 caractères)")
		}
		icon := strings.TrimSpace(it.Icon)
		if icon == "" {
			icon = "fa-wifi"
		}
		if !portalServiceIcons[icon] {
			return "", fmt.Errorf("icône non supportée : %s", icon)
		}
		out = append(out, service{Icon: icon, Label: label})
	}
	if len(out) == 0 {
		return "", nil // liste vidée = section « Nos Services » masquée
	}
	b, err := json.Marshal(out)
	if err != nil {
		return "", err
	}
	return string(b), nil
}

// portalEncodeSlides — carrousel commercial (N°136) : ≤ 3 URLs https ≤ 300
// car. Liste vide = retour aux 3 images génériques du template.
func portalEncodeSlides(list []string) (string, error) {
	if len(list) > 3 {
		return "", fmt.Errorf("Slides invalides : au plus 3 slides")
	}
	out := make([]string, 0, len(list))
	for _, it := range list {
		u := strings.TrimSpace(it)
		if u == "" {
			continue // entrée vide ignorée (la console n'en envoie pas)
		}
		if !strings.HasPrefix(u, "https://") {
			return "", fmt.Errorf("Slides invalides : URL https:// requise")
		}
		if len(u) > 300 {
			return "", fmt.Errorf("Slides invalides : URL trop longue")
		}
		out = append(out, u)
	}
	if len(out) > 0 {
		b, err := json.Marshal(out)
		if err != nil {
			return "", fmt.Errorf("Slides invalides : %v", err)
		}
		return string(b), nil
	}
	return "", nil // liste vidée = slides retirées → images par défaut
}

// portalEncodeTicker — bandeau animé (N°138) : ≤ 5 messages de 1-80 car.
// (texte brut trimé). Liste vide = retour aux 3 messages par défaut.
func portalEncodeTicker(list []string) (string, error) {
	if len(list) > 5 {
		return "", fmt.Errorf("Messages du bandeau invalides : au plus 5 messages")
	}
	out := make([]string, 0, len(list))
	for _, it := range list {
		m := strings.TrimSpace(it)
		if m == "" {
			continue
		}
		if len(m) > 80 {
			return "", fmt.Errorf("Messages du bandeau invalides : 80 caractères max par message")
		}
		out = append(out, m)
	}
	if len(out) > 0 {
		b, err := json.Marshal(out)
		if err != nil {
			return "", fmt.Errorf("Messages du bandeau invalides : %v", err)
		}
		return string(b), nil
	}
	return "", nil
}

// portalEncodeWhatsapp — numéro WhatsApp SUPPORT (N°139) : 8-15 chiffres en
// format international (espaces/+/-/() tolérés puis retirés), label ≤ 30.
// Number vide = numéro retiré → repli support MikCloud.
func portalEncodeWhatsapp(req portalWhatsappReq) (string, error) {
	num := hotpage.WhatsappNumber(req.Number)
	if req.Number != "" && num == "" {
		return "", fmt.Errorf("Numéro WhatsApp invalide : 8 à 15 chiffres en format international requis (ex. 2250708091012)")
	}
	label := strings.TrimSpace(req.Label)
	if len(label) > 30 {
		return "", fmt.Errorf("Libellé WhatsApp trop long (30 caractères max)")
	}
	if num == "" {
		return "", nil
	}
	type wa struct {
		Number string `json:"number"`
		Label  string `json:"label,omitempty"`
	}
	b, err := json.Marshal(wa{Number: num, Label: label})
	if err != nil {
		return "", fmt.Errorf("Numéro WhatsApp invalide : %v", err)
	}
	return string(b), nil
}

// assemblePortalOverride — valide CHAQUE champ de la requête d'une surcharge
// (mêmes bornes que les réglages du compte) et construit le JSON CANONIQUE
// (model.CanonicalPortalOverride : champs vides omis, "" si surcharge vide).
// L'appelant stocke la valeur telle quelle : elle rejoint le miroir de
// signature hotspot_files sans re-sérialisation supplémentaire.
func assemblePortalOverride(req *portalOverrideReq) (string, error) {
	if req == nil {
		return "", nil
	}
	var ov model.PortalOverride
	if req.DisplayName != nil {
		v := strings.TrimSpace(*req.DisplayName)
		if len(v) > 80 {
			return "", fmt.Errorf("nom affiché trop long (80 caractères max)")
		}
		ov.DisplayName = v
	}
	if req.LogoURL != nil {
		v := strings.TrimSpace(*req.LogoURL)
		if err := portalValidateLogo(v); err != nil {
			return "", err
		}
		ov.LogoURL = v
	}
	if req.BannerURL != nil {
		v := strings.TrimSpace(*req.BannerURL)
		if err := portalValidateBanner(v); err != nil {
			return "", err
		}
		ov.BannerURL = v
	}
	if req.WaveLink != nil {
		v := strings.TrimSpace(*req.WaveLink)
		if len(v) > 300 {
			return "", fmt.Errorf("lien marchand Wave trop long (300 caractères max)")
		}
		ov.WaveLink = v
	}
	if req.Style != nil {
		v := strings.TrimSpace(*req.Style)
		if err := portalValidateStyle(v); err != nil {
			return "", err
		}
		ov.Style = v
	}
	if req.Welcome != nil {
		v := strings.TrimSpace(*req.Welcome)
		if err := portalValidateWelcome(v); err != nil {
			return "", err
		}
		ov.Welcome = v
	}
	if req.Promos != nil {
		s, err := portalEncodePromos(*req.Promos)
		if err != nil {
			return "", err
		}
		if s != "" {
			// La string JSON persistée a les MÊMES json tags que
			// model.PortalPromoLine : le décodage est homéomorphe.
			_ = json.Unmarshal([]byte(s), &ov.Promos)
		}
	}
	if req.Socials != nil {
		s, err := portalEncodeSocials(*req.Socials)
		if err != nil {
			return "", err
		}
		if s != "" {
			_ = json.Unmarshal([]byte(s), &ov.Socials)
		}
	}
	if req.Services != nil {
		s, err := portalEncodeServices(*req.Services)
		if err != nil {
			return "", err
		}
		if s != "" {
			_ = json.Unmarshal([]byte(s), &ov.Services)
		}
	}
	if req.Slides != nil {
		s, err := portalEncodeSlides(*req.Slides)
		if err != nil {
			return "", err
		}
		if s != "" {
			_ = json.Unmarshal([]byte(s), &ov.Slides)
		}
	}
	if req.Ticker != nil {
		s, err := portalEncodeTicker(*req.Ticker)
		if err != nil {
			return "", err
		}
		if s != "" {
			_ = json.Unmarshal([]byte(s), &ov.Ticker)
		}
	}
	if req.Whatsapp != nil {
		s, err := portalEncodeWhatsapp(*req.Whatsapp)
		if err != nil {
			return "", err
		}
		if s != "" {
			_ = json.Unmarshal([]byte(s), &ov.Whatsapp)
		}
	}
	return model.CanonicalPortalOverride(ov), nil
}

File: frontend/src/lib/hotspot/portal-branding.ts
// N°186 — domaine PARTAGÉ du branding du portail captif : la forme d'édition
// (13 champs), les décodeurs défensifs, la résolution de la chaîne
// ROUTEUR → SITE → COMPTE et la construction du corps de surcharge.
//
// Miroir EXACT de la sémantique backend (internal/api/portal_branding.go et
// resolvePortalBranding/applyPortalOverride de portal_serve.go) :
//   • « VIDE = HÉRITE » — un champ vide ne touche rien, une liste vide non
//     plus, un numéro WhatsApp vide non plus ;
//   • la surcharge est REMPLACÉE EN ENTIER à chaque écriture — l'éditeur
//     envoie donc toujours l'état complet souhaité (groupes réinitialisés =
//     champs omis du corps) ;
//   • les bornes de validation sont partagées avec les réglages du compte.
//
// Consommé par le formulaire du compte (hotspot-cards — HotspotForm, via les
// décodeurs et la forme structurelle PortalBrandingFields) et par l'éditeur
// unifié à sélecteur de contexte N°186 (portal-editor — surcharges
// SITE/ROUTEUR) : les MÊMES règles de lecture et d'écriture aux trois
// niveaux, zéro divergence console ↔ serveur.

import type {
  AppSettings,
  PortalOverrideData,
  PortalPromo,
  PortalService,
  PortalSocial,
} from "@/lib/hotspot/types";

/* ─── Formes d'édition ─── */

/** PortalBrandingFields — les 10 champs de branding édités par les briques
 * PARTAGÉES du formulaire riche (bannière, carrousel, services, bandeau,
 * WhatsApp, mode d'affichage). Forme STRUCTURELLE : le HotspotForm du compte
 * (hotspot-cards) et le PortalOverrideForm de l'éditeur N°186 la satisfont
 * tous deux — les briques se rendent ainsi À L'IDENTIQUE aux trois niveaux
 * de la chaîne. */
export interface PortalBrandingFields {
  bannerUrl: string;
  slides: string[];
  services: PortalService[];
  ticker: string[];
  waNumber: string;
  waLabel: string;
  portalStyle: string;
  welcome: string;
  promos: PortalPromo[];
  socials: PortalSocial[];
}

/** PortalOverrideForm — TOUT l'état éditable d'une personnalisation de
 * portail : les 10 champs de branding + l'identité propre au niveau (nom
 * affiché, logo, lien marchand Wave). Vide = hérite. */
export interface PortalOverrideForm extends PortalBrandingFields {
  displayName: string;
  logoUrl: string;
  waveLink: string;
}

/* ─── Groupes de l'éditeur unifié ─── */

/** Groupes pédagogiques de l'éditeur : un groupe hérite ou est personnalisé
 * EN BLOC. « personnalisé » = au moins un de ses champs porteur de valeur
 * (même prédicat que la sémantique « vide = hérite » du backend). */
export type OverrideGroupKey =
  | "identity"
  | "banner"
  | "slides"
  | "services"
  | "ticker"
  | "whatsapp"
  | "display";

export const PORTAL_OVERRIDE_GROUPS: readonly OverrideGroupKey[] = [
  "identity",
  "banner",
  "slides",
  "services",
  "ticker",
  "whatsapp",
  "display",
];

/* ─── Décodeurs défensifs (JSON invalide ou absent = valeur neutre) ─── */
/* Déplacés de hotspot-cards (N°186) : ils servent désormais AUSSI à décoder
 * les valeurs héritées du compte dans l'éditeur de surcharge. Comportement
 * conservé À L'IDENTIQUE. */

export function parseStringArray(raw: string | undefined): string[] {
  try {
    const parsed = JSON.parse(raw || "[]") as string[];
    return Array.isArray(parsed) ? parsed.filter((u) => typeof u === "string" && u) : [];
  } catch {
    return [];
  }
}

export function parseServices(raw: string | undefined): PortalService[] {
  try {
    const parsed = JSON.parse(raw || "[]") as PortalService[];
    return Array.isArray(parsed) ? parsed : [];
  } catch {
    return [];
  }
}

export function parseWhatsapp(raw: string | undefined): { number: string; label: string } {
  try {
    const parsed = JSON.parse(raw || "{}") as { number?: unknown; label?: unknown };
    return {
      number: typeof parsed.number === "string" ? parsed.number : "",
      label: typeof parsed.label === "string" ? parsed.label : "",
    };
  } catch {
    return { number: "", label: "" };
  }
}

export function parsePromos(raw: string | undefined): PortalPromo[] {
  try {
    const parsed = JSON.parse(raw || "[]") as PortalPromo[];
    return Array.isArray(parsed) ? parsed : [];
  } catch {
    return [];
  }
}

export function parseSocials(raw: string | undefined): PortalSocial[] {
  try {
    const parsed = JSON.parse(raw || "[]") as PortalSocial[];
    return Array.isArray(parsed) ? parsed : [];
  } catch {
    return [];
  }
}

/* ─── Résolution de la chaîne (miroir de resolvePortalBranding) ─── */

/** emptyBrandingForm — forme vierge : tout hérite, rien n'est personnalisé. */
export function emptyBrandingForm(): PortalOverrideForm {
  return {
    displayName: "",
    logoUrl: "",
    bannerUrl: "",
    waveLink: "",
    portalStyle: "",
    welcome: "",
    slides: [],
    services: [],
    ticker: [],
    promos: [],
    socials: [],
    waNumber: "",
    waLabel: "",
  };
}

/** accountBrandingOf — le branding EFFECTIF du compte (base de la chaîne) :
 * les réglages du tenant décodés dans la forme d'édition. Même lecture
 * défensive que l'ancien initialForm de hotspot-cards (champ absent du JSON =
 * défaut effectif, pas « vide »). */
export function accountBrandingOf(settings: AppSettings): PortalOverrideForm {
  const wa = parseWhatsapp(settings.tenant.portalWhatsapp);
  return {
    displayName: settings.tenant.name ?? "",
    logoUrl: settings.tenant.logoUrl ?? "",
    bannerUrl: settings.tenant.bannerUrl ?? "",
    waveLink: settings.tenant.waveLink ?? "",
    portalStyle: settings.tenant.portalStyle ?? "",
    welcome: settings.tenant.portalWelcome ?? "",
    slides: parseStringArray(settings.tenant.portalSlides),
    services: parseServices(settings.tenant.portalServices),
    ticker: parseStringArray(settings.tenant.portalTicker),
    promos: parsePromos(settings.tenant.portalPromos),
    socials: parseSocials(settings.tenant.portalSocials),
    waNumber: wa.number,
    waLabel: wa.label,
  };
}

/** applyOverrideForm — applique une surcharge décodée sur un branding de
 * base, champ par champ : seules les valeurs NON VIDES gagnent (miroir
 * exact d'applyPortalOverride — la surcharge ne peut pas masquer un élément
 * que le niveau inférieur affiche). */
export function applyOverrideForm(
  base: PortalOverrideForm,
  ov: PortalOverrideData,
): PortalOverrideForm {
  return {
    ...base,
    displayName: ov.displayName ? ov.displayName : base.displayName,
    logoUrl: ov.logoUrl ? ov.logoUrl : base.logoUrl,
    bannerUrl: ov.bannerUrl ? ov.bannerUrl : base.bannerUrl,
    waveLink: ov.waveLink ? ov.waveLink : base.waveLink,
    portalStyle: ov.portalStyle ? ov.portalStyle : base.portalStyle,
    welcome: ov.portalWelcome ? ov.portalWelcome : base.welcome,
    slides: ov.portalSlides && ov.portalSlides.length > 0 ? [...ov.portalSlides] : base.slides,
    services:
      ov.portalServices && ov.portalServices.length > 0
        ? ov.portalServices.map((s) => ({ icon: s.icon ?? "fa-wifi", label: s.label }))
        : base.services,
    ticker: ov.portalTicker && ov.portalTicker.length > 0 ? [...ov.portalTicker] : base.ticker,
    promos:
      ov.portalPromos && ov.portalPromos.length > 0
        ? ov.portalPromos.map((p) => ({
            id: p.id,
            title: p.title,
            desc: p.desc ?? "",
            imageUrl: p.imageUrl ?? "",
            priceLabel: p.priceLabel ?? "",
            link: p.link,
          }))
        : base.promos,
    socials:
      ov.portalSocials && ov.portalSocials.length > 0
        ? ov.portalSocials.map((s) => ({ label: s.label, url: s.url }))
        : base.socials,
    waNumber: ov.portalWhatsapp?.number ? ov.portalWhatsapp.number : base.waNumber,
    waLabel: ov.portalWhatsapp?.number ? (ov.portalWhatsapp.label ?? "") : base.waLabel,
  };
}

/** formFromOverride — état initial de l'éditeur pour une surcharge
 * persistée (JSON canonique) : les champs absents restent vides (= hérite). */
export function formFromOverride(ov: PortalOverrideData): PortalOverrideForm {
  return applyOverrideForm(emptyBrandingForm(), ov);
}

/* ─── Prédicats par groupe ─── */

/** groupIsSet — le groupe porte-t-il une personnalification ? Même
 * sémantique que le backend : au moins un champ porteur de valeur (compte
 * tenu des trim/normalisations d'encodage). */
export function groupIsSet(f: PortalOverrideForm, key: OverrideGroupKey): boolean {
  switch (key) {
    case "identity":
      return f.displayName.trim() !== "" || f.logoUrl.trim() !== "" || f.waveLink.trim() !== "";
    case "banner":
      return f.bannerUrl.trim() !== "";
    case "slides":
      return f.slides.some((u) => u.trim() !== "");
    case "services":
      return f.services.some((s) => s.label.trim() !== "");
    case "ticker":
      return f.ticker.some((m) => m.trim() !== "");
    case "whatsapp":
      return f.waNumber.replace(/[^0-9]/g, "") !== "";
    case "display":
      return (
        f.portalStyle !== "" ||
        f.welcome.trim() !== "" ||
        f.promos.some((p) => p.title.trim() !== "") ||
        f.socials.some((s) => s.label.trim() !== "" || s.url.trim() !== "")
      );
  }
}

const sameJson = (a: unknown, b: unknown) => JSON.stringify(a) === JSON.stringify(b);

/** groupEquals — compare les champs D'UN groupe entre deux états (compteur
 * « modifié » et barre d'action N°140 de l'éditeur). */
export function groupEquals(a: PortalOverrideForm, b: PortalOverrideForm, key: OverrideGroupKey): boolean {
  switch (key) {
    case "identity":
      return a.displayName === b.displayName && a.logoUrl === b.logoUrl && a.waveLink === b.waveLink;
    case "banner":
      return a.bannerUrl === b.bannerUrl;
    case "slides":
      return sameJson(a.slides, b.slides);
    case "services":
      return sameJson(a.services, b.services);
    case "ticker":
      return sameJson(a.ticker, b.ticker);
    case "whatsapp":
      return a.waNumber === b.waNumber && a.waLabel === b.waLabel;
    case "display":
      return (
        a.portalStyle === b.portalStyle &&
        a.welcome === b.welcome &&
        sameJson(a.promos, b.promos) &&
        sameJson(a.socials, b.socials)
      );
  }
}

/* ─── Manipulation de groupes (personnaliser / réinitialiser) ─── */

/** groupValuesOf — les champs D'UN groupe, extraits d'une forme source (pour
 * initialiser la personnalisation depuis la valeur HÉRITÉE : l'utilisateur
 * part de ce qui est servi aujourd'hui et ajuste). */
export function groupValuesOf(f: PortalOverrideForm, key: OverrideGroupKey): Partial<PortalOverrideForm> {
  switch (key) {
    case "identity":
      return { displayName: f.displayName, logoUrl: f.logoUrl, waveLink: f.waveLink };
    case "banner":
      return { bannerUrl: f.bannerUrl };
    case "slides":
      return { slides: [...f.slides] };
    case "services":
      return { services: f.services.map((s) => ({ ...s })) };
    case "ticker":
      return { ticker: [...f.ticker] };
    case "whatsapp":
      return { waNumber: f.waNumber, waLabel: f.waLabel };
    case "display":
      return {
        portalStyle: f.portalStyle,
        welcome: f.welcome,
        promos: f.promos.map((p) => ({ ...p })),
        socials: f.socials.map((s) => ({ ...s })),
      };
  }
}

/** emptyGroupValuesOf — les champs d'un groupe remis à vide (= hérite à
 * nouveau, le groupe disparaît du corps envoyé). */
export function emptyGroupValuesOf(key: OverrideGroupKey): Partial<PortalOverrideForm> {
  switch (key) {
    case "identity":
      return { displayName: "", logoUrl: "", waveLink: "" };
    case "banner":
      return { bannerUrl: "" };
    case "slides":
      return { slides: [] };
    case "services":
      return { services: [] };
    case "ticker":
      return { ticker: [] };
    case "whatsapp":
      return { waNumber: "", waLabel: "" };
    case "display":
      return { portalStyle: "", welcome: "", promos: [], socials: [] };
  }
}

/* ─── Corps de surcharge (contrat PUT /api/sites/{id} et
     PUT /api/routers/{id}/portal — remplacement ENTIER à chaque écriture,
     champs vides omis = hérite) ─── */

/** overridePayloadOf — construit le corps de surcharge correspondant à
 * l'état de l'éditeur : seuls les champs porteurs de valeur partent (miroir
 * enrichi de l'ancien formToPayload : promos, réseaux et icônes de services
 * inclus — la parité de champs du formulaire riche s'applique aux trois
 * niveaux). */
export function overridePayloadOf(f: PortalOverrideForm): Record<string, unknown> {
  const payload: Record<string, unknown> = {};
  const displayName = f.displayName.trim();
  if (displayName) payload.displayName = displayName;
  const logoUrl = f.logoUrl.trim();
  if (logoUrl) payload.logoUrl = logoUrl;
  const bannerUrl = f.bannerUrl.trim();
  if (bannerUrl) payload.bannerUrl = bannerUrl;
  const waveLink = f.waveLink.trim();
  if (waveLink) payload.waveLink = waveLink;
  if (f.portalStyle) payload.portalStyle = f.portalStyle;
  const welcome = f.welcome.trim();
  if (welcome) payload.portalWelcome = welcome;
  const slides = f.slides.map((u) => u.trim()).filter((u) => u !== "");
  if (slides.length > 0) payload.portalSlides = slides;
  const services = f.services.filter((s) => s.label.trim() !== "");
  if (services.length > 0) payload.portalServices = services;
  const ticker = f.ticker.map((m) => m.trim()).filter((m) => m !== "");
  if (ticker.length > 0) payload.portalTicker = ticker;
  const promos = f.promos.filter((p) => p.title.trim() !== "");
  if (promos.length > 0) payload.portalPromos = promos;
  const socials = f.socials.filter((s) => s.label.trim() !== "" || s.url.trim() !== "");
  if (socials.length > 0) payload.portalSocials = socials;
  const waDigits = f.waNumber.replace(/[^0-9]/g, "");
  if (waDigits) payload.portalWhatsapp = { number: f.waNumber.trim(), label: f.waLabel.trim() };
  return payload;
}
